Make ConcurrentResourceClaimSyncs configurable

anson627 opened the pull request:
This pull request introduces configuration options for the ResourceClaim controller in the Kubernetes controller manager, enabling users to control the number of concurrent resource claim syncs. It adds support for flag parsing, validation, application of options, and defaulting logic for the ResourceClaim controller. Comprehensive unit tests are also included to ensure correct behavior.

These changes make the ResourceClaim controller configurable and ensure its options are validated and tested, improving flexibility and robustness.

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
make resource claim concurrency configurable for large training job run with 10k pods on 10k nodes

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
add --concurrent-resourceclaim-syncs to configure kube-controller-manager resource claim reconcile concurrency
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

mortent commented:
This seems to follow similar patterns as other controllers. Is this PR due to a concrete need to make concurrency configurable?

pohly commented:
/retitle WIP: Make ConcurrentResourceClaimSyncs configurable

anson627 commented:
> I can see how this might be useful, but currently this looks like it will need further work before the code is ready for review (see test failures).
> 
> Out of curiosity, did you use AI tools to generate the code?

Thanks for checking! I will fix the test failures.

no, I don't feel like AI can handle large complex code base like kubernetes yet

pohly committed:
DRA resource claim controller: configurable number of workers

It might never be necessary to change the default, but it is hard to be sure.
It's better to have the option, just in case.

anson627 commented:
Hey @pohly, I fixed almost all CI checks, and PR is ready for review again, I also updated the PR description to make it more human readable, copilot is too chatty : )

Regarding the 2 failed CI checks:

[pull-kubernetes-integration](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/134701/pull-kubernetes-integration/1983887216564244480) is not related, also I saw it flapping from time to time

[pull-kubernetes-verify](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/134701/pull-kubernetes-verify/1983887221668712448) need a bit guidance and help from you, I tried to run `hack/verify-codegen.sh`, but found it make following changes which caused more CI checks to fail
```
diff --git a/pkg/controller/apis/config/v1alpha1/zz_generated.conversion.go b/pkg/controller/apis/config/v1alpha1/zz_generated.conversion.go
index 6cb4e2daeb9..290211ad3c4 100644
--- a/pkg/controller/apis/config/v1alpha1/zz_generated.conversion.go
+++ b/pkg/controller/apis/config/v1alpha1/zz_generated.conversion.go
@@ -46,7 +46,7 @@ import (
 	podgcconfigv1alpha1 "k8s.io/kubernetes/pkg/controller/podgc/config/v1alpha1"
-	if err := resourceclaimconfigv1alpha1.Convert_config_ResourceClaimControllerConfiguration_To_v1alpha1_ResourceClaimControllerConfiguration(&in.ResourceClaimController, &out.ResourceClaimController, s); err != nil {
+	if err := Convert_config_ResourceClaimControllerConfiguration_To_v1alpha1_ResourceClaimControllerConfiguration(&in.ResourceClaimController, &out.ResourceClaimController, s); err != nil {
 		return err
 	}
 	if err := resourcequotaconfigv1alpha1.Convert_config_ResourceQuotaControllerConfiguration_To_v1alpha1_ResourceQuotaControllerConfiguration(&in.ResourceQuotaController, &out.ResourceQuotaController, s); err != nil {
diff --git a/pkg/controller/resourceclaim/config/v1alpha1/zz_generated.conversion.go b/pkg/controller/resourceclaim/config/v1alpha1/zz_generated.conversion.go
deleted file mode 100644
index dfb0ea0c79c..00000000000
--- a/pkg/controller/resourceclaim/config/v1alpha1/zz_generated.conversion.go
+++ /dev/null
```

pohly reviewed (commented):
Looks good to me with one small nit.

anson627 commented:
> Well, it _looks_ good to poor human, but the machines disagree: code generation is still not fixed.

fixed and pushed again
```
./hack/verify-codegen.sh 
installing goimports from hack/tools
Downloading https://github.com/protocolbuffers/protobuf/releases/download/v23.4/protoc-23.4-linux-x86_64.zip succeed
Archive:  protoc-v23.4-linux-amd64.zip
  inflating: protoc-v23.4-linux-amd64/bin/protoc  
  inflating: protoc-v23.4-linux-amd64/include/google/protobuf/any.proto  
  inflating: protoc-v23.4-linux-amd64/include/google/protobuf/api.proto  
  inflating: protoc-v23.4-linux-amd64/include/google/protobuf/compiler/plugin.proto  
  inflating: protoc-v23.4-linux-amd64/include/google/protobuf/descriptor.proto  
  inflating: protoc-v23.4-linux-amd64/include/google/protobuf/duration.proto  
  inflating: protoc-v23.4-linux-amd64/include/google/protobuf/empty.proto  
  inflating: protoc-v23.4-linux-amd64/include/google/protobuf/field_mask.proto  
  inflating: protoc-v23.4-linux-amd64/include/google/protobuf/source_context.proto  
  inflating: protoc-v23.4-linux-amd64/include/google/protobuf/struct.proto  
  inflating: protoc-v23.4-linux-amd64/include/google/protobuf/timestamp.proto  
  inflating: protoc-v23.4-linux-amd64/include/google/protobuf/type.proto  
  inflating: protoc-v23.4-linux-amd64/include/google/protobuf/wrappers.proto  
  inflating: protoc-v23.4-linux-amd64/readme.txt  
protoc v23.4 installed. To use:
export PATH="/tmp/verify-generated-update-codegen.sh.VXmwBW/third_party/protoc:${PATH}"
+++ [1104 20:42:09] Generating protobufs for 72 targets
+++ [1104 20:42:32] Generating deepcopy code for 264 targets
+++ [1104 20:42:38] Generating swagger for 60 targets
+++ [1104 20:42:43] Generating prerelease-lifecycle code for 61 targets
+++ [1104 20:42:46] Generating defaulter code for 105 targets
+++ [1104 20:42:52] Generating validation code for 123 targets
+++ [1104 20:42:55] Generating conversion code for 148 targets
+++ [1104 20:43:01] Generating register code for 1 targets
+++ [1104 20:43:07] Generating openapi code
+++ [1104 20:43:27] Generating apply-config code for 61 targets
+++ [1104 20:43:35] Generating client code for 55 targets
+++ [1104 20:43:40] Generating lister code for 60 targets
+++ [1104 20:43:42] Generating informer code for 60 targets
+++ [1104 20:43:44] Generating code for subproject staging/src/k8s.io/code-generator/examples
    Generating deepcopy code for 15 targets
    Generating defaulter code for 11 targets
    Generating conversion code for 4 targets
    Generating register code for 15 targets
    Generating openapi code for 4 targets
    Generating client code for 4 targets
    Generating lister code for 4 targets
    Generating informer code for 4 targets
    Generating applyconfig code for 4 targets
    Generating client code for 4 targets
    Generating lister code for 4 targets
    Generating informer code for 4 targets
    Generating applyconfig code for 1 targets
    Generating client code for 1 targets
    Generating lister code for 1 targets
    Generating informer code for 1 targets
    Generating applyconfig code for 1 targets
    Generating client code for 1 targets
    Generating lister code for 1 targets
    Generating informer code for 1 targets
    Generating applyconfig code for 1 targets
    Generating client code for 1 targets
    Generating lister code for 1 targets
    Generating informer code for 1 targets
+++ [1104 20:44:07] Generating code for subproject staging/src/k8s.io/kube-aggregator
    Generating deepcopy code for 3 targets
    Generating defaulter code for 2 targets
    Generating conversion code for 2 targets
    Generating openapi code for 2 targets
    Generating client code for 2 targets
    Generating lister code for 2 targets
    Generating informer code for 2 targets
+++ [1104 20:44:14] Generating code for subproject staging/src/k8s.io/sample-apiserver
    Generating deepcopy code for 3 targets
    Generating defaulter code for 2 targets
    Generating conversion code for 2 targets
    Generating openapi code for 2 targets
    Generating applyconfig code for 2 targets
    Generating client code for 2 targets
    Generating lister code for 2 targets
    Generating informer code for 2 targets
+++ [1104 20:44:22] Generating code for subproject staging/src/k8s.io/sample-controller
    Generating deepcopy code for 1 targets
    Generating client code for 1 targets
    Generating lister code for 1 targets
    Generating informer code for 1 targets
+++ [1104 20:44:25] Generating code for subproject staging/src/k8s.io/metrics
    Generating deepcopy code for 8 targets
    Generating conversion code for 5 targets
    Generating client code for 2 targets
+++ [1104 20:44:30] Generating code for subproject staging/src/k8s.io/apiextensions-apiserver
    Generating deepcopy code for 4 targets
    Generating defaulter code for 2 targets
    Generating conversion code for 2 targets
    Generating openapi code for 3 targets
    Generating applyconfig code for 2 targets
    Generating client code for 2 targets
    Generating lister code for 2 targets
    Generating informer code for 2 targets
+++ [1104 20:44:39] Generating code for subproject staging/src/k8s.io/apiextensions-apiserver/examples/client-go
    Generating deepcopy code for 1 targets
    Generating applyconfig code for 1 targets
    Generating client code for 1 targets
    Generating lister code for 1 targets
    Generating informer code for 1 targets
+++ [1104 20:44:43] Generating protobuf bindings for 10 targets
```

anson627 commented:
everything else passed now except [pull-kubernetes-dra-integration](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/134701/pull-kubernetes-dra-integration/1985809876965658624), I think it's not related, but I can re-run if there is concern

anson627 committed:
generate files

pohly commented:
/test pull-kubernetes-node-e2e-containerd

https://github.com/kubernetes/kubernetes/issues/135132

pohly commented:
That's a known issue unrelated to your PR. Let's retest once we have LGTM/approval.

soltysh reviewed (approved):
/lgtm
/approve

There's a minor nit, that I'll leave a hold so you can fix it before landing this. But feel free to drop the hold. 

/hold

anson627 committed:
resourceclaimcontroller: normalize validation error message

pohly commented on cmd/kube-controller-manager/app/options/resourceclaimcontroller.go:
```diff
@@ -0,0 +1,63 @@
+
+	var errs []error
+	if o.ConcurrentSyncs <= 0 {
+		errs = append(errs, fmt.Errorf("concurrent-resource-claim-syncs must be larger than zero, got %d", o.ConcurrentSyncs))
```
No hurry, we are in code freeze anyway.

/lgtm

pohly commented:
/test pull-kubernetes-integration

https://github.com/kubernetes/kubernetes/issues/135540

liggitt commented on cmd/kube-controller-manager/app/options/resourceclaimcontroller.go:
```diff
@@ -0,0 +1,63 @@
+		return
+	}
+
+	fs.Int32Var(&o.ConcurrentSyncs, "concurrent-resource-claim-syncs", o.ConcurrentSyncs, "The number of operations (creating or deleting ResourceClaims) allowed to run concurrently. Larger number = more responsive, but more CPU (and network) load")
```
nit: be consistent with the canonical controller name

```
	ResourceClaimController                            = "resourceclaim-controller"
```

```suggestion
	fs.Int32Var(&o.ConcurrentSyncs, "concurrent-resourceclaim-syncs", o.ConcurrentSyncs, "The number of operations (creating or deleting ResourceClaims) allowed to run concurrently. Larger number = more responsive, but more CPU (and network) load")
```

anson627 committed:
Update cmd/kube-controller-manager/app/options/resourceclaimcontroller.go

Co-authored-by: Jordan Liggitt <jordan@liggitt.net>

soltysh reviewed (approved):
The changed Jordan requested was applied.
/lgtm
/hold cancel

liggitt commented:
/approve
